app/src/sharedTest/java/org/oppia/app/onboarding/OnboardingFragmentTest.kt

```diff
@@ -0,0 +1,486 @@
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.espresso.matcher.ViewMatchers.withTagValue
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.hamcrest.CoreMatchers.allOf
+import org.hamcrest.CoreMatchers.equalTo
+import org.hamcrest.CoreMatchers.not
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.app.R
+import org.oppia.app.profile.ProfileActivity
+
+/** Tests for [OnboardingFragment]. */
+@RunWith(AndroidJUnit4::class)
+class OnboardingFragmentTest {
+
+  @Before
+  fun setUp() {
+    Intents.init()
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlideTitle_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(
+        allOf(
+          withId(R.id.slide_title_text_view),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withText(R.string.slide_0_title)))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlideDescription_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(
+        allOf(
+          withId(R.id.slide_description_text_view),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withText(R.string.slide_0_description)))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlideImage_isCorrect() {
+    launch(OnboardingActivity::class.java).use {
+      onView(
+        allOf(
+          withId(R.id.slide_image_view),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withTagValue(equalTo(R.drawable.ic_onboarding_0))))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlide_index0DotIsActive_otherDotsAreInactive() {
+    launch(OnboardingActivity::class.java).use {
+      onView(
+        allOf(
+          withId(R.id.onboarding_dot_0),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withAlpha(1.0F)))
+      onView(
+        allOf(
+          withId(R.id.onboarding_dot_1),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withAlpha(0.3F)))
+      onView(
+        allOf(
+          withId(R.id.onboarding_dot_2),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withAlpha(0.3F)))
+      onView(
+        allOf(
+          withId(R.id.onboarding_dot_3),
+          isCompletelyDisplayed()
+        )
+      ).check(matches(withAlpha(0.3F)))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlide_skipButtonIsVisible() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.skip_text_view)).check(matches(isDisplayed()))
+    }
+  }
+
+  @Test
+  fun testOnboardingFragment_checkDefaultSlide_clickSkipButton_shiftsToLastSlide() {
+    launch(OnboardingActivity::class.java).use {
+      onView(withId(R.id.skip_text_view)).perform(click())
+      SystemClock.sleep(500)
```

Comment:
Here and elsewhere: we should use our custom waitFor() instead of a hard sleep. Sleeps like this are fragile and slow down the tests.

Reply:
Minimal code with only one test case which uses waitFor() code, currently this test case is failing but in mobile is working as we are expecting it to be.

https://gist.github.com/rt4914/d4793eb6e7be36615e75966e14c36ff3

Reply:
Two major points on how all these test cases have been fixed:

1. As per @BenHenning 's suggestion I replaced `swipe` with `ViewPagerActions.swipeToPage()` and that fixed some of the issues. `swipeToPage()` waits for everything to finish before going to the next step in test case.

2. Using `show-layout-bounds` from developer options in mobile, I was able to see some boxes which were actually meant to be `Visibility.GONE`, so that was also fixed, but changing the margin to padding.